### PR TITLE
Add feature that creates book orders.

### PR DESCRIPTION
**Pull request by ak47na**
`OrderHandler.java` class will be used by the server to make orders via `makeOrders(String userId, Point address, List<String> bookIds).`
A future PR will implement the methods in `BookManager.java.`

### CODE REVIEW

**Review by anamariamilcu — APPROVED** [review]

File: src/main/java/com/google/sps/data/GoogleMapsPathFinder.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+package com.google.sps.data;
+
+import com.google.maps.errors.ApiException;
+import com.google.sps.data.MapsRequest;
+import java.io.IOException;
+import java.lang.InterruptedException;
+
+public class GoogleMapsPathFinder implements PathFinder {
+  public int getTimeBetweenPoints(Point A, Point B) throws ApiException, IOException, InterruptedException {
```
Comment (ghost) [inline src/main/java/com/google/sps/data/GoogleMapsPathFinder.java]: Nit: use `lowerCamelCase` for parameter names. https://google.github.io/styleguide/javaguide.html#s5.2.6-parameter-names

```suggestion
  public int getTimeBetweenPoints(Point a, Point a) throws ApiException, IOException, InterruptedException { 
```

File: src/main/java/com/google/sps/data/LibraryPoint.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  public LibraryPoint(double latitude, double longitude, int id) {
+    super(latitude, longitude);
+    booksToBorrow = new ArrayList<>();
+    this.libraryId = id;
+  }
+
+  public int getLibraryId() {
+    return libraryId;
+  }
+  public void setLibraryId(int libraryId) {
```
Comment (ghost) [inline src/main/java/com/google/sps/data/LibraryPoint.java]: Missing space between functions

```suggestion
  }
  
  public void setLibraryId(int libraryId) {
```

File: src/main/java/com/google/sps/data/LibraryPoint.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    this.libraryId = libraryId;
+  }
+
+  /**
+   * Adds the Google Books ids of the books from bookIdsString to the books that must be borrowed
+   * from this library during current delivery
+   */
+  public void addBooks(String bookIdsString) {
```
Comment (ghost) [inline src/main/java/com/google/sps/data/LibraryPoint.java]: No need to put the type in the name of the variable.

```suggestion
  public void addBooks(String bookIds) {
```

File: src/main/java/com/google/sps/data/LibraryPoint.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  public int getLibraryId() {
+    return libraryId;
+  }
+  public void setLibraryId(int libraryId) {
+    this.libraryId = libraryId;
+  }
+
+  /**
+   * Adds the Google Books ids of the books from bookIdsString to the books that must be borrowed
+   * from this library during current delivery
```
Comment (ghost) [inline src/main/java/com/google/sps/data/LibraryPoint.java]: Can you add a sentence on this to indicate what the `bookIds` string might look like, e.g.

`The booksId should be in the format 'bookId1&bookId2&bookID3'` (...if that indeed is the forma you are expectingt?!)

File: src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+package com.google.sps.data;
+
+public class ManhattanDistancePathFinder implements PathFinder {
+  public int manhattanDistance(Point A, Point B) {
+    return (int)(Math.abs(A.latitude - B.latitude) + Math.abs(A.longitude - B.longitude));
+  }
+
+  public int getTimeBetweenPoints(Point A, Point B) { 
+    return manhattanDistance(A, B);
```
Comment (ghost) [inline src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java]: Nit: lowerCamelCase as before.  A->a and B->b

File: src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+package com.google.sps.data;
+
+public class ManhattanDistancePathFinder implements PathFinder {
+  public int manhattanDistance(Point A, Point B) {
```
Comment (ghost) [inline src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java]: Since this class is already called `ManhattanDistancePathFinder`, it is probably redundant to call this `manhattanDistance` as well right?  `new ManhattanDistancePathFinder().manhattanDistance(x,y)` seems a bit like we are repeating ourselves?

Could it be more concise to just call this `distance` so you end up with something like `new ManhattanDistancePathFinder().distance(x,y)`?

File: src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    return (int)(Math.abs(A.latitude - B.latitude) + Math.abs(A.longitude - B.longitude));
```
Comment (ghost) [inline src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java]: Does this really need to return an `int`?  The `int` distance between a latitude and longitude is quite high in real-world terms right?

File: src/main/java/com/google/sps/data/MapsRequest.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import com.google.maps.DirectionsApi;
+import com.google.maps.errors.ApiException;
+import com.google.maps.GeoApiContext;
+import com.google.maps.GeoApiContext.Builder;
+import com.google.maps.GeocodingApi;
+import com.google.maps.model.*;
+import java.io.IOException;
+
+/**
+ * Class that create a GeoApiContext variable and handles requests to maps API.
```
Comment (ghost) [inline src/main/java/com/google/sps/data/MapsRequest.java]: Nit: typo

```suggestion
 * Class that creates a GeoApiContext variable and handles requests to maps API.
```

File: src/main/java/com/google/sps/data/MapsRequest.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+ */
+public class MapsRequest {
+  private static GeoApiContext geoApiContext = null;
+  public static GeoApiContext getGeoApiContext() {
+    if (geoApiContext == null) {
+      geoApiContext = new GeoApiContext
+          .Builder()
+          .apiKey("")
```
Comment (ghost) [inline src/main/java/com/google/sps/data/MapsRequest.java]: Should this have a value?

Might be worth leaving a `TODO(ak47na): provide Maps API key` so it is not forgotten?

File: src/main/java/com/google/sps/data/MapsRequest.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  /**
+   * Makes a request to Geocoding API and returns the latitude and longitude associated with the
+   * address String.
+   */
+  public static LatLng getLocationFromAddress(String address) throws IOException, ApiException, InterruptedException {
+    GeocodingResult[] results =  GeocodingApi.geocode(getGeoApiContext(), address).await();
+    Gson gson = new GsonBuilder().setPrettyPrinting().create();
+    return results[0].geometry.location;
```
Comment (ghost) [inline src/main/java/com/google/sps/data/MapsRequest.java:55]: What happens if results.length == 0?

Might be worth checking for that here and throwing an exception if it is not what you are expecting.

File: src/main/java/com/google/sps/data/MapsRequest.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  /**
+   * Makes a request to Geocoding API and returns the duration in seconds to get from point A to point B
+   * TODO[ak47na]: update function to get the duration if the journey starts at a given time.
+   * TODO[ak47na]: add function parameter to support multiple transportation modes.
+   */
+  static int getTimeBetween2Points(Point A, Point B) throws ApiException, IOException, InterruptedException{
```
Comment (ghost) [inline src/main/java/com/google/sps/data/MapsRequest.java]: Nit: lowerCamelCase

```suggestion
  static int getTimeBetween2Points(Point a, Point b) throws ApiException, IOException, InterruptedException {
```

File: src/main/java/com/google/sps/data/MapsRequest.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    DirectionsResult result =
+        DirectionsApi.newRequest(getGeoApiContext())
+            .origin(new LatLng(A.latitude, A.longitude))
+            .destination(new LatLng(B.latitude, B.longitude))
+            .mode(TravelMode.DRIVING)
+            .await();
+    int timeInSeconds = 0;
+    for (DirectionsLeg leg : result.routes[0].legs) {
```
Comment (ghost) [inline src/main/java/com/google/sps/data/MapsRequest.java:75]: As before, what happens if the length of this array is 0?

File: src/main/java/com/google/sps/data/MapsRequest.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+      timeInSeconds += leg.duration.inSeconds;
+    }
+    return timeInSeconds;
```
Comment (ghost) [inline src/main/java/com/google/sps/data/MapsRequest.java:78]: If the length of the array is zero, you'll return a `timeInSeconds` of 0 here right?  That is "correct" in terms of your code (it wont crash), but that strikes me as a bug in the logic - unless both A & B are the exact same latlng this should never happen right?

If that is the case then it might be worth throwing an exception here so that calling code knows something went wrong.

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+package com.google.sps.data;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.maps.errors.ApiException;
+import java.io.IOException;
+import java.lang.InterruptedException;
+import java.lang.*;
```
Comment (ghost) [inline src/main/java/com/google/sps/data/OrderHandler.java]: Please avoid wildcrd imports. https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

File: src/main/java/com/google/sps/data/PathFinder.java [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.maps.errors.ApiException;
+import java.io.IOException;
+import java.lang.InterruptedException;
+
+interface PathFinder {
+  public int getTimeBetweenPoints(Point A, Point B)  throws ApiException, IOException, InterruptedException ;
```
Comment (ghost) [inline src/main/java/com/google/sps/data/PathFinder.java]: Nit: lowerCamelCase

```suggestion
  public int getTimeBetweenPoints(Point a, Point b)  throws ApiException, IOException, InterruptedException ;
```

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+   * library that has all the requested books.
+   */
+  private void addOrderToDatastore(LibraryPoint library, ArrayList<String> bookIds, String userId, Point address) {
+    Entity order = new Entity("Order");
+    order.setProperty("libraryId", (int)library.getLibraryId());
+    order.setProperty("libraryLatitude", library.latitude);
+    order.setProperty("libraryLongitude", library.longitude);
+    order.setProperty("recipientLatitude", address.latitude);
+    order.setProperty("recipientLongitude", address.longitude);
+    order.setProperty("status", "ADDED");
```
Comment (ghost) [inline src/main/java/com/google/sps/data/OrderHandler.java]: Optional: It might make sense to have the statuses as an `enum` somewhere to avoid "magic strings" all over the place.

Nice example of how to add string values to enums https://www.baeldung.com/java-enum-values

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    order.setProperty("area", library.getArea());
+    order.setProperty("userId", userId);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(order);
+  }
+
+  /**
+   * Given a user's Id, the shipping address and an array of book ids, it creates a set of orders
+   * and adds them to datastore.
```
Comment (ghost) [inline src/main/java/com/google/sps/data/OrderHandler.java]: Can you add something about what value is returned and why? E.g. when is `false` returned

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+   */
+  public boolean makeOrders(String userId, Point address, List<String> bookIds) throws ApiException, IOException, InterruptedException {
+    HashMap <LibraryPoint, ArrayList<String>> libraryBookIds = new HashMap<>();
+
+    for (String bookId : bookIds) {
+      Entity closestLibrary = getClosestLibrary(
+          BooksManager.getLibrariesForBook(bookId, /** setLimit = */false), address);
+      if (closestLibrary == null) {
+        // there is no library that has bookId, thus the order can't be made
+        return false;
```
Comment (ghost) [inline src/main/java/com/google/sps/data/OrderHandler.java]: So if I had an order for say 10 books, and just one cannot be made, the whole thing is rejected.

That is fine, but how does the calling coded know which book was rejected?

Could it make more sense to return a collection back to the user so that we have a true/false status for each individual book, with an overall status for the entire order?  You might need to make an `OrderStatus` class to return that info.

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+      }
+      LibraryPoint library = new LibraryPoint((double) closestLibrary.getProperty("libraryLatitude"),
```
Comment (ghost) [inline src/main/java/com/google/sps/data/OrderHandler.java]: Can you move these magic strings to a static constant somewhere?  Will help avoid mistakes in the future if someone mis-types

File: src/main/java/com/google/sps/data/Point.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+  /**
+   * Returns the area of the point from its latitude and longitude.
+   * TODO[ak47na]: change function to return different area values depending on coordinates.
+   */
+  public int getArea() {
+    return 1;
+  }
+
+  /**
+   * Returns true if .this Point has latitude and longitude equal to the ones of B
```
Comment (ghost) [inline src/main/java/com/google/sps/data/Point.java]: Nit: Don't need to know about implementation's private variable names :-)

```suggestion
   * Returns true if this Point has equal latitude and longitude
```

File: src/main/java/com/google/sps/data/Point.java [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+   */
+  @Override
+  public boolean equals(Object object) {
+    if (!(object instanceof Point)) {
+      return false;
+    }
+    Point B = (Point) object;
```
Comment (ghost) [inline src/main/java/com/google/sps/data/Point.java]: Nit: lowerCamelCase

**Review by ghost — CHANGES_REQUESTED** [review]
Looks good - please add some tests e.g. here are some things that look like good candidates to test:

- Point.java: equals and hashcode
- OrderHandler.java: makeOrders and getClosestLibrary
- ManHattanDistancePathFinder.java: manhattanDistance

Remember what we mentioned before - look at the boundary cases for the values you are providing, particularly when things are zero-length.

File: src/main/java/com/google/sps/data/OrderHandler.java (hunk 19 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/OrderHandler.java]: Yes, I added them as constants.

File: src/main/java/com/google/sps/data/OrderHandler.java (hunk 18 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/OrderHandler.java]: You're right, I changed the function to return a Collection with the book ids of the rejected books. I can change it to return a Collection of `OrderStatus` objects if that is better, but the status of books can be inferred knowing which books are rejected.

File: src/main/java/com/google/sps/data/OrderHandler.java (hunk 16 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/OrderHandler.java]: You're right, I added an `OrderStatus` enum.

File: src/main/java/com/google/sps/data/MapsRequest.java (hunk 13 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/MapsRequest.java:78]: Thank you for noticing, I changed the function to throw DataNotFoundException in that case.

File: src/main/java/com/google/sps/data/MapsRequest.java (hunk 10 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/MapsRequest.java:55]: I changed it to throw DataNotFoundException if `results == null || results.length == 0`.

File: src/main/java/com/google/sps/data/MapsRequest.java (hunk 9 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/MapsRequest.java]: Yes, I added it as a todo.

File: src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java (hunk 7 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java]: `ManhattanDistancePathFinder` implements `PathFinder` interface and it's only used for testing. I use int distance because `GoogleMapsPathFinder.getTimeBetweenPoints(a, b) `, the PathFinder used in production computes the `int` number of seconds to get from point a to point b.

Is there a better way to mock `GoogleMapsPathFinder.getTimeBetweenPoints(a, b) `?

File: src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java (hunk 6 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java]: Right, I changed the function name to distance.

File: src/main/java/com/google/sps/data/LibraryPoint.java (hunk 4 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/LibraryPoint.java]: Yes, I added a comment specifying the format.

File: src/main/java/com/google/sps/data/LibraryPoint.java (hunk 3 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/LibraryPoint.java]: I changed the name to `bookIds`.

File: src/main/java/com/google/sps/data/LibraryPoint.java (hunk 2 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/LibraryPoint.java]: I added the space.

File: src/main/java/com/google/sps/data/GoogleMapsPathFinder.java (hunk 1 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/GoogleMapsPathFinder.java]: I changed the parameter names.

File: src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java (hunk 7 above)
Comment (ghost) [reply src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java]: I see.

If this function is just used to calculate the "cost" of one route vs another, it might be worth keeping the precision of floating point lat longs and just multiplying them by say 100000 to get a larger int that keeps the precision.

So e.g. the lat lng of two of Google's London offices 6PS and CSG:
* 6PS 51.5206431,-0.135168
* CSG 51.5134121,-0.1407977

So we end up with (51.5206431 - 51.5134121) + (-0.135168 - -0.1407977) = 0.012860700000004416 which as an int is 0 manhattan distance but they are ~2km apart straight-line distance in the real world

If we multiplied those lat longs up before doing the we'd keep some of the precision:
* 6PS 5152064.31,−13516.8,
* CSG 5151341.21,-140797.7

This gives us a more "realistic" manhattan distance/cost of 1286, and your routing will be able to better distinguish things close together in an urban environment.

Up to you - I guess it depends how close all of the points will be in your application.

File: src/main/java/com/google/sps/data/OrderHandler.java (hunk 18 above)
Comment (ghost) [reply src/main/java/com/google/sps/data/OrderHandler.java]: A list of rejected books sounds good to me.

**Review by ghost — COMMENTED** [review]
Thanks for the changes - I'm not seeing some of the changes yet (E.g. the Point A, B -> Point a, b) but I guess that is just not pushed yet.

File: src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java (hunk 7 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/ManhattanDistancePathFinder.java]: You're right, it makes a lot of sense. I changed the distance function to return the value multiplied by 100000 and updated the tests.

**Review by ghost — APPROVED** [review]
